### ISSUE TITLE
Feature: BeforeFeature / AfterFeature hooks

### ISSUE THREAD

**Issue by vitalets** [vitalets/playwright-bdd#219]
**The problem**
Sometimes we need to run code before / after particular feature file. Currently it can be achieved in playwright-bdd only with some [hacks](https://github.com/vitalets/playwright-bdd/issues/216#issuecomment-2345646232). 
In CucumberJS it's not possible - there were several related requests:
- https://github.com/cucumber/cucumber-js/issues/914 
- https://github.com/cucumber/cucumber-js/issues/1284
- https://github.com/cucumber/cucumber-js/issues/1447
- https://github.com/cucumber/messages/issues/260

CucumberJS FAQ states that these hooks [are not possible](https://github.com/cucumber/cucumber-js/blob/main/docs/faq.md#why-arent-there-beforefeature-and-afterfeature-hooks), because runner does not consider features when running scenarios.

**A solution**
Playwright provides a direct way to run code before particular file - [test.beforeAll / test.afterAll](https://playwright.dev/docs/api/class-test#test-before-all). We can just utilize that.
SpecFlow also [provides](https://docs.specflow.org/projects/specflow/en/latest/Bindings/Hooks.html#supported-hook-attributes) BeforeFeature / AfterFeature hooks.

Binding to features can be done by tags - the same as for scenario hooks. Example code:
```ts
const { BeforeFeature, AfterFeature } = createBdd();

BeforeFeature({ tags: '@auth' }, async () => { 
  // run setup before features with @auth tag 
});
```

This also matches the proposed [tags from path approach](https://github.com/vitalets/playwright-bdd/issues/217).
You can put feature files and BeforeFeature hooks into `@`-prefixed directory, and they will be bound automatically:
```
features
└── @auth
    ├── signin.feature
    └── hooks.ts
```

**Hooks naming**
Hooks naming between Cucumber and Playwright is a bit confusing:
* Playwright's `BeforeEach / AfterEach` = Cucumber's `Before / After` - run before/after each scenario
* Playwright's `BeforeAll / AfterAll` = (no alternative in CucumberJS) - run before/after each feature
* Playwright's worker-scoped auto-fixtures = Cucumber's `BeforeAll / AfterAll` - run before/after each worker

I like the explicit naming in SpecFlow:
* `BeforeScenario / AfterScenario`
* `BeforeFeature / AfterFeature`
* `BeforeWorker / AfterWorker`

I think, we can use these aliases in playwright-bdd.

**Running once**
There is a request from community to run `beforeFeature` hooks only once, even in `fullyParallel` mode / in case when some tests are failing and new worker is created.
Keep in mind during implementation, see https://github.com/microsoft/playwright/issues/22520#issuecomment-2351421682 

I you have any thoughts / concerns here, feel free to share in comments.
Or just add your 👍

**Comment by vitalets** [comment]
**What should be a default behavior of `BeforeFeature / AfterFeature` hooks?**

1. Run once in **each worker**: this is how Playwright's `test.beforeAll / test.afterAll` currently behave. If some test fails, new worker is created and hooks re-run again.
2. Run once in a **whole test run**: this is not how Playwright's `test.beforeAll / test.afterAll` behave, but according to [playwright#22520](https://github.com/microsoft/playwright/issues/22520) it is often needed (to setup some database stuff, etc).

What do you think?

**Comment by alescinskis** [comment]
I vote for 1st option

**Comment by vitalets** [comment]
> What should be a default behavior of BeforeFeature / AfterFeature hooks?

I thought more about that question. If `BeforeFeature` runs in each worker, it's actually not `BeforeFeature` but `BeforeWorker`. A confusion goes from naming: while feature is related to BDD level, worker is related to runner. Mixing these names in hooks does not make it clean.

Finally, I propose the following changes for hooks:
1. Alias existing `Before / After` hooks as `BeforeScenario / AfterScenario` and recommend using the latter, as they are more explicit
2. Alias existing `BeforeAll / AfterAll` as `BeforeWorker / AfterWorker` and recommend using the latter, as they are more explicit
3. Allow to pass `tags` option to `BeforeWorker / AfterWorker`. If `tags` are passed, the hook runs only if matched features are being executed (so actually behaves like BeforeFeature / AfterFeature). If a new worker is created for matched feature, the hook runs again. If `tags` are not passed, the hook runs before / after each worker (current behavior).
4. Allow to pass `once` option to `BeforeWorker / AfterWorker`. If `once: true` is passed, the hook runs only once during the whole test run session (so actually behave like Playwright's global setup / teardown). If additionally filtered by `tags`, the hook runs once and only if matched features are being executed - this is the main benefit for setting up some data once for the particular feature.
5. No `BeforeFeature / AfterFeature` hooks

Example code:
```ts
const { BeforeWorker } = createBdd();

BeforeWorker({ tags: '@auth', once: true }, async () => { 
  // heavy setup of test user in database for checking all auth stuff
});
```
Feature file can contain `@auth` tag on the top level:
```feature
@auth
Feature: check auth

  Scenario: check login
```
Or on the scenario level:
```feature
Feature: check auth

  @auth
  Scenario: check login
```
If at least one scenario of feature has the matching tag, `BeforeWorker` hook will run before that feature.

**Comment by vitalets** [comment]
Playwright-BDD v8 [shipped](https://vitalets.github.io/playwright-bdd/#/blog/whats-new-in-v8?id=tagged-beforeall-and-afterall) tagging support for `BeforeAll` / `AfterAll`. So the following requirement is fulfilled:

> Allow to pass `tags` option to BeforeWorker / AfterWorker. If tags are passed, the hook runs only if matched features are being executed (so actually behaves like BeforeFeature / AfterFeature). If a new worker is created for matched feature, **the hook runs again**. If tags are not passed, the hook runs before / after each worker (current behavior).

So, `BeforeFeature / AfterFeature` can be now implemented with tags:
```ts
BeforeAll({ tags: '@my-feature' }, async () => { ... });
AfterAll({ tags: '@my-feature' }, async () => { ... });
```

One note here, that `BeforeFeature / AfterFeature` can be interpreted in different ways. 
If there are two feature files with `@my-feature` tag, should `BeforeAll` run before each feature or only once?
The same for parallelization: if a feature is split across several workers, should `BeforeAll` in each worker or not?

In current implementation, tagged worker hooks will run in each worker - similar to Playwright beforeAll / afterAll hooks.

But in some cases this is not a desirable behavior. If I want to populate database for some feature, I'd like to do it once, no matter how many workers run my tests. This is a goal for mentioned `once` option, it's not implemented now. There are some ideas, how to achieve that, but it's not an easy task, will be investigated separately.

Closing for now. Feel free to open a new issue if needed.
